File: gcc/testsuite/gfortran.dg/gomp/dispatch-1.f90
module main
  use iso_c_binding, only: c_ptr
  implicit none
  contains
  
  subroutine f1 ()
    integer :: a, b, arr(10)
    real :: x
    complex :: c
    character :: ch
    logical :: bool
    type :: struct
      integer :: a
      real :: b
    end type
    type(struct) :: s
    type(c_ptr) :: p
    
    interface
    subroutine f0 (a, c, bool, s)
      import :: struct
      integer, intent(in) :: a
      complex, intent(out) :: c
      logical, intent(inout) :: bool
      type(struct) :: s
    end subroutine
    integer function f2 (arr, x, ch, b)
      integer, intent(inout) :: arr(:)
      real, intent(in) :: x
      character, intent(out) :: ch
      real :: b
    end function
    subroutine f3 (p)
      import :: c_ptr
      type(c_ptr) :: p
    end subroutine
    integer function f4 ()
    end function
    end interface

    !$omp dispatch
      b = f2(arr, x, ch, s%b)
    !$omp dispatch
      c = f2(arr(:5), x * 2.4, ch, s%b)
    !$omp dispatch
      arr(1) = f2(arr, x, ch, s%b)
    !$omp dispatch
      s%a = f2(arr, x, ch, s%b)
    !$omp dispatch
      x = f2(arr, x, ch, s%b)
    !$omp dispatch
      call f0(a, c, bool, s)
    !$omp dispatch
      call f0(f4(), c, bool, s)
      
    !$omp dispatch nocontext(.TRUE.)
      call f0(a, c, bool, s)
    !$omp dispatch nocontext(arr(2) < 10)
      call f0(a, c, bool, s)
    !$omp dispatch novariants(.FALSE.)
      call f0(a, c, bool, s)
    !$omp dispatch novariants(bool)
      call f0(a, c, bool, s)
    !$omp dispatch nowait
      call f0(a, c, bool, s)
    !$omp dispatch device(arr(9))
      call f0(a, c, bool, s)
    !$omp dispatch device(a + a)
      call f0(a, c, bool, s)
    !$omp dispatch device(-25373654)
      call f0(a, c, bool, s)
    !$omp dispatch is_device_ptr(p)
      call f3(p)
    !$omp dispatch depend(in: a, c, bool) depend(inout: s, arr(:3))
      call f0(a, c, bool, s)
  end subroutine
end module

File: gcc/testsuite/gcc.dg/gomp/dispatch-1.c
/* Test parsing of #pragma omp dispatch */
/* { dg-do compile } */

int f0 (int);

void f1 (void)
{
  int a, b;
  double x;
  struct {int a; int b;} s;
  int arr[1];

#pragma omp dispatch
  int c = f0 (a);	/* { dg-error "expected expression before 'int'" } */
#pragma omp dispatch
  int f2 (int d);	/* { dg-error "expected expression before 'int'" } */
#pragma omp dispatch
  a = b;	/* { dg-error "expected a function name before ';' token" } */
#pragma omp dispatch
  s.a = f0(a) + b;	/* { dg-error "expected ';' before '\\+' token" } */
#pragma omp dispatch
  b = !f0(a);	/* { dg-error "expected a function name before '!' token" } */
#pragma omp dispatch
  s.b += f0(s.a);	/* { dg-error "expected '=' before '\\+=' token" } */
#pragma omp dispatch
#pragma omp threadprivate(a)	/* { dg-error "expected expression before '#pragma'" } */
  a = f0(b);
  
#pragma omp dispatch nocontext(s) /* { dg-error "used struct type value where scalar is required" } */
  f0(a);
#pragma omp dispatch nocontext(a, b) /* { dg-error "expected '\\)' before ','" } */
  f0(a);
#pragma omp dispatch nocontext(a) nocontext(b) /* { dg-error "too many 'nocontext' clauses" } */
  f0(a);
#pragma omp dispatch novariants(s) /* { dg-error "used struct type value where scalar is required" } */
  f0(a);
#pragma omp dispatch novariants(a, b) /* { dg-error "expected '\\)' before ','" } */
  f0(a);
#pragma omp dispatch novariants(a) novariants(b) /* { dg-error "too many 'novariants' clauses" } */
  f0(a);
#pragma omp dispatch nowait nowait /* { dg-error "too many 'nowait' clauses" } */
  f0(a);
#pragma omp dispatch device(x) /* { dg-error "expected integer expression before end of line" } */
  f0(a);
#pragma omp dispatch device(arr) /* { dg-error "expected integer expression before end of line" } */
  f0(a);
#pragma omp dispatch is_device_ptr(x) /* { dg-error "'is_device_ptr' variable is neither a pointer nor an array" } */
  f0(a);
#pragma omp dispatch is_device_ptr(&x) /* { dg-error "expected identifier before '&' token" } */
  f0(a);
#pragma omp dispatch depend(inout: f0) /* { dg-error "'f0' is not lvalue expression nor array section in 'depend' clause" } */
  f0(a);
}
